### DISCUSSION TITLE
Support the Review Panel in Overleaf

### DISCUSSION THREAD

**Discussion by truthlive** [overleaf-workshop/Overleaf-Workshop#64] [Ideas]
When working on the journal paper with multiple collaborators on Overleaf, they usually provide many comments directly on the Review Panel of Overleaf. I have to open the Overleaf website separately to view and address their comments, which is not convenient. 

Would it be possible to fetch the comments from collaborators in the Review Panel and view them in VScode? It would be better if we could directly address their comments using the extension but viewing these comments should be enough for me. 

I suppose the edit type of changes (such as add or delete text) in the Review Panel can be ignored for now since they may not be easily displayed in VScode.

**Comment by iamhyc** [comment]
The feature "Review Panel" is only provided in "Server Pro Edition" and the official Overleaf website, which implies there is no public API documentation or source code for reference. There will be a huge workload, and I am not sure whether it deserves or not.

Anyway, I will firstly note down my ideas below. Let's firstly discuss how much we can achieve in vscode!

### API Reverse Engineering
**Show Threads List**:
- get all threads: `GET project/{projectId}/threads`
- display resolved threads: `GET project/{projectId}/ranges`
- delete resolved thread: `DELETE project/{projectId}/doc/{docId}/thread/{threadId}`
  - deleted event emit: websocket `delete-thread`
- reopen resolved thread: `POST project/{projectId}/thread/{threadId}/reopen`
  - reopened event emit: websocket `reopen-thread`

**Comment in Thread**:
- new/reply comment: `POST project/{projectId}/thread/{threadId}/messages`
  - update event emit: websocket `new-comment` with `{"c":<original-text-commented-on>, "p":<posotion-offset>, "t":<thread-id>}`
- edit thread: `POST project/{projectId}/thread/{threadId}/messages/{messageId}/edit`
  - updated event emit: websocket `edit-message`
- resolve thread: `POST project/{projectId}/thread/{threadId}/resolve`
  - resolved event emit: websocket `resolve-thread`

**Track Changes**:
- track add/delete by adding `meta:{tc: <tc-id>}` in `applyOtUpdate` in websocket
  > as no source code available, we have to guess how to generate "\<tc-id\>" locally
- all tracked changes got when `joinDoc` in websocket
  > implies maybe no RESTful API available to extract/differ the "tc" update
- accept/reject via `applyOtUpdate` in websocket, i.e., normal editing in the document

### How to Display
- How to display a column with custom view contained besides the document? I got no idea now.
- To mark the Commented/Added/Deleted section in document, it is easy: just wrap them in `setDecorations`

**Reply by iamhyc** [reply]
I think we can crack the API thing, as shown in the notes above, there should be no other hidden APIs.

The major obstacle is the design: we can't have a side-by-side review panel in vscode. One possible solution come to my mind is as below.
![](https://github.com/iamhyc/Overleaf-Workshop/assets/9068301/bc5c824b-a730-4a0e-a5ab-1706f8324e3d)

**Reply by truthlive** [reply]
Hi @iamhyc,

Thank you so much for the detailed information! I apologize that I don't know much about the ShareLatex API and have no idea this would be such a huge workload for you. I also have no experience in API reverse engineering but your ideas seem very promising to me.

Regarding the display of comments, in my case, my collaborators usually turn off "track changes" and only provide comments for certain highlighted sentences. And we usually do not use "reply" to discuss in the review panel. So my initial thought was that we only fetch these comments to VScode and display them in the "problems" panel of VScode, similar to a spelling check tool that shows both incorrect words and errors (see below figure). This would be convenient to quickly view and address the comments in the latex file which is enough for me.
![image](https://github.com/iamhyc/Overleaf-Workshop/assets/25575959/2d926180-e5c1-4e05-8442-d9d0e3dbf64a)

I do think that your design of the side-by-side panel is much better and more useful than mine, especially with all the functions from Overleaf's review panel supported. But it also means much more work for the implementation, right :sweat:? I agree with you that the function may not deserve implementation since we are not sure how many people will need it. Maybe we can wait for more input and ideas from the community. Anyway, thanks again for your response!

**Reply by iamhyc** [reply]
You don't need to apologize for anything, it is a good idea itself. I am glad that people would like to use this extension with more useful features.
Actually, the so-called "reverse engineering" is just to observe the format the HTTP/Web Socket packets appeared while interacting with the Overleaf Web editor. It is easy to do so with the help of browser built-in developer tools.

As for the display of comments, I think have the comments listed in problem panel is good but not enough. You may have too many annoying "squiggly lines" in same color in the document context, and cannot tell what the "squiggly lines" stand for at first look. That is why I think there at least should be "an icon displayed besides".

As for the review panel, a quick implementation would be like the "spell check feature": display the comments in hover message, and reply/resolve via quick actions (reply in a popup input box)

![](https://github.com/iamhyc/Overleaf-Workshop/assets/9068301/95bd4b26-8237-4818-96c9-084d477d4e8b)

**Reply by truthlive** [reply]
Thanks for the explanation! 

For the "API reverse engineering", I remember some unofficial ChatGPT libs used a similar way to get the prompt's answer but the accounts used in these libs sometimes got limited. Since the "review panel" is in the Overleaf Server Pro Edition, I guess we should be careful not to have too many calls for interacting with the review panel, right? Not sure if the account with a subscription would be fine or not.

Regarding the display of comments, I agree with you that the problem panel idea is pretty rough and does not have a good visualization to identify the comments. An icon would be very helpful in this case.

The quick implementation you mentioned seems good👍! If more people find this function useful or need this feature, it can always be extended with a better design in the future😄.

**Comment by juandpinto** [comment]
I have nothing to contribute here except to say that I think having access to collaborators' comments would be a huge help. It's really the only feature that makes it so that I still need to open Overleaf on the browser.

**Comment by iamhyc** [comment]
I have put this alone for more comments so far. Since I have some free time now, I can now start to work on a rough version.

@truthlive Hi, if you have time, please help create a formal "Feature Request" issue [here](https://github.com/iamhyc/Overleaf-Workshop/issues/new?assignees=&labels=feature+request&projects=&template=feature_request.md&title=%5BFeature+Request%5D). Thanks for the help!

**Reply by truthlive** [reply]
Hi @iamhyc I create the "feature request" issue here: https://github.com/iamhyc/Overleaf-Workshop/issues/94
Let me know if these is anything I need to edit in the issue. Thank you for working on this feature!!

**Reply by iamhyc** [reply]
Thank you! The information is clear enough in the PR.

**Comment by iamhyc** [comment]
Hi, I have just verified the function of review comments related operations. You can have a try with the temporary extension [here](https://github.com/iamhyc/Overleaf-Workshop/actions/runs/7948023956).

![图片](https://github.com/iamhyc/Overleaf-Workshop/assets/9068301/0a5cc560-2aec-4aa6-8801-11a69b963ee7)

The UI is still rough but I am not going to improve it now. I will continue to implement the "track change" function and then see what we can do with the UI.

**Reply by truthlive** [reply]
Hi @iamhyc, thanks for the update!👍I just tried the temporary extension, and the review comments-related operations work well for me. It is so convenient to directly check the comments of the highlighted sentences within the latex file!

I agree with you that the UI can be improved (maybe with an overview of all comments as in Overleaf?) but should not be the development priority now.

**Comment by iamhyc** [comment]
Hi guys, the track change thing is far more complex than I expected. Luckily, I do find Overleaf release the source code of related code in their repo, e.g., https://github.com/overleaf/overleaf/tree/main/libraries/ranges-tracker.
~~Now I need more time to investigate if it is possible and legal to integrate their source code.~~ Well, it turns out I just need to *steal* the seed generator function :)

So far, we only provide the "Review Comment" feature in the temporary extension [overleaf-workshop-0.9.2-0432dd3.vsix](https://github.com/iamhyc/Overleaf-Workshop/actions/runs/7948023956).
If anyone is interested in the development progress, please keep track of the latest change in this PR #95

**Reply by JKargo** [reply]
Hi @iamhyc, great work!
I would really like to use this feature. Unfortunately the download has expired.

**Reply by iamhyc** [reply]
Hi, you can find the latest temporary extension in the list. I hope the review comments function is not broken now.

https://github.com/iamhyc/Overleaf-Workshop/actions/workflows/vsce-package.yml?query=branch%3A94-review-panel

As for the "track change" function, I currently do not have sufficient time to work on it and my premium trial is expired now.
I may pick it up next week.

**Comment by Musicminion** [comment]
Hello! Review panel is not some feature difficult to imply, and Now, I can provide a naive implements of overleaf pro.

Release:
https://github.com/ayaka-notes/overleaf/tree/097cfb0386cdbabbde22336f588c682912531735


Effect display:
https://blog.ayaka.space/2024/02/Overleaf-Pro-Effect-Display/

Hopefully this can be helpful to your development.

**Comment by SunJianier** [comment]
I would really like to use this feature. Unfortunately the download has expired. How can I get this feature?

**Reply by SunJianier** [reply]
I will have installed why is the review still not displayed automatically. Are there any additional operations that need to be done.

**Reply by iamhyc** [reply]
Hi, I have re-run the workflow, and you can try this temporary extension: https://github.com/iamhyc/Overleaf-Workshop/actions/runs/13615834960

**Reply by LpldW** [reply]
Hi, I came across this discussion about review panel and the expired Actions artifact for the VSIX build. I’m very interested in trying it out.
Is there any newer VSIX virg this feature available for testing?
Also, is (or was) the feature stable enough for trial use?
Many thanks!
